feat: add tic-tac-toe React prototype

zairahira opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #642

<!-- Feel free to add any additional description of changes below this line -->

jdwilkin4 commented:
Hey @zairahira !

The prototype looks good so far.

This will be discussed more in tuesday's meeting, but we are changing what will be shipped by Christmas.
We are going to ship everything from HTML through React fundamentals. This change will be more manageable for all teams involved 👍 

As a result, let's push pause on the tic toe game and resume in Q1 of 2025

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/index.jsx:
```diff
@@ -0,0 +1,84 @@
+export const Square = ({ value, onClick }) => {
```
looking at the current react implementation on learn, I don't know if we are going to be able to export multiple child components into a parent component. 

right now, it looks like we just have one `index.jsx` file and the parent and child components are all in that file.
So we might need to remove these export keywords for all of the child components.

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/index.jsx:
```diff
@@ -0,0 +1,84 @@
+  );
+};
+
+export const calculateWinner = (squares) => {
```
see earlier note about exports

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/index.jsx:
```diff
@@ -0,0 +1,84 @@
+  return null;
+};
+
+export const Board = () => {
```
see earlier note about exports

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/user-stories.md:
```diff
@@ -0,0 +1,14 @@
+1. You should create and export a `Square` component that accepts two props: `value` and `onClick`.
```
we will probably need to remove all user stories pointing to export child components into a parent component file because I don't think our current system is setup for that.

looking at the current projects, there is only one `index.jsx` file and both parent and child components live there
https://www.freecodecamp.org/learn/full-stack-developer/lab-mood-board/build-a-mood-board

For now, lets remove all user stories about exporting child components. 

```suggestion
1. You should create a `Square` component that accepts two props: `value` and `onClick`.
```

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/user-stories.md:
```diff
@@ -0,0 +1,14 @@
+5. You should create and export a `calculateWinner` function that accepts an array `squares` as a parameter and returns the winner if there is one.
+6. Your `calculateWinner` function should check for winning lines using an array of indices for rows, columns, and diagonals, and return the value of the winner (`X` or `O`) if there is a match.
+7. You should create and export a `Board` component.
+8. Your `Board` component should use React's `useState` to keep track of the game state, including the `squares` array and whether it is player `X` or `O` turn.
```
```suggestion
8. Your `Board` component should use React's `useState` hook to keep track of the game state, including the `squares` array and whether it is player `X` or `O` turn.
```

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/user-stories.md:
```diff
@@ -0,0 +1,14 @@
+9. Your `Board` component should call the `calculateWinner` function to check if there's a winner after each move.
+10. Your `Board` component should render a status message that shows the current player or the winner, or indicate if there is a draw.
+11. Your `Board` component should render a grid of `Square` components for the 3x3 Tic-Tac-Toe board.
+12. You should create and export a `resetGame` function that resets the game state to its initial values when triggered.
```
```suggestion
12. You should create a `resetGame` function that resets the game state to its initial values when triggered.
```

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/user-stories.md:
```diff
@@ -0,0 +1,14 @@
+10. Your `Board` component should render a status message that shows the current player or the winner, or indicate if there is a draw.
+11. Your `Board` component should render a grid of `Square` components for the 3x3 Tic-Tac-Toe board.
+12. You should create and export a `resetGame` function that resets the game state to its initial values when triggered.
+13. You should create and export an `App` component that renders the `Board` component.
```
based on the current react projects here 

https://www.freecodecamp.org/learn/full-stack-developer/lab-mood-board/build-a-mood-board

it looks like we will need to export just the `Board` component from the `index.jsx` file. so we can probably remove App altogether and refactor it so it looks like the other projects on `learn`

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/user-stories.md:
```diff
@@ -0,0 +1,14 @@
+11. Your `Board` component should render a grid of `Square` components for the 3x3 Tic-Tac-Toe board.
+12. You should create and export a `resetGame` function that resets the game state to its initial values when triggered.
+13. You should create and export an `App` component that renders the `Board` component.
+14. The `App` component should be rendered to the DOM inside a `div` with the class `App`.
```
we could remove this because looking at the other react projects, it looks like we are already giving them the boilerplate code in the `index.html`

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/index.jsx:
```diff
@@ -0,0 +1,84 @@
+  );
+};
+
+export const App = () => {
+  return (
+    <div className="App">
+      <Board />
+    </div>
+  );
+};
```
it looks like we can remove this and just export the `Board` instead based on the current project on `learn`

zairahira commented:
I have implemented the following suggestions:
- Exported the `Board` component instead of multiple components.
- Updated the `index.html` to properly import and render the `Board` component.
- Remove `App`.
- Update user stories.

jdwilkin4 commented on fullstack-cert/react-projects/tic-tac-toe/user-stories.md:
```diff
@@ -0,0 +1,12 @@
+2. Your `Square` component should return a `button` element with the class `square`.
+3. The `button` element should display the `value` prop inside it.
+4. The `Square` component should trigger the `onClick` function when the button is clicked.
+5. You should create and export a `calculateWinner` function that accepts an array `squares` as a parameter and returns the winner if there is one.
```
this function won't be exported

```suggestion
5. You should create a `calculateWinner` function that accepts an array `squares` as a parameter and returns the winner if there is one.
```
